src/bootstrap/builder.rs

```diff
@@ -869,21 +869,24 @@ impl<'a> Builder<'a> {
         self.try_run(patchelf.arg(fname));
     }
 
-    pub(crate) fn download_component(
-        &self,
-        base: &str,
-        url: &str,
-        dest_path: &Path,
-        help_on_error: &str,
-    ) {
+    pub(crate) fn download_component(&self, url: &str, dest_path: &Path, help_on_error: &str) {
         // Use a temporary file in case we crash while downloading, to avoid a corrupt download in cache/.
         let tempfile = self.tempdir().join(dest_path.file_name().unwrap());
+        // While bootstrap itself only supports http and https downloads, downstream forks might
+        // need to download components from other protocols. The match allows them adding more
+        // protocols without worrying about merge conficts if we change the HTTP implementation.
+        match url.split_once("://").map(|(proto, _)| proto) {
```

Comment:
Maybe move this match into `download_with_retries`?

Reply:
Hmm, is the move to `download_with_retries` just for code organization, or because you'd like retries to be handled regardless of the protocol?

In the latter case, that'd require a larger refactor of `download_with_retries`:

* Curl retries are not handled in Rust code, but through the `--retries` flag. We'd need to move that into Rust code.
* HTTP first tries curl, and then on Windows it tries PowerShell. Encoding that in a generic retry mechanism would be slightly annoying.

Reply:
I assumed `download_with_retries` would be called from other places, but it only has a single callsite. Keeping the match where it currently is is fine.